File: dsa/linked-list/linked-list.js
// Node constructor
class Node {
    constructor(value) {
        this.value = value
        this.next = null
    }
}

// Linked List constructor of size 1
class LinkedList {
    constructor(value) {
        let new_node = new Node(value)
        this.head = new_node
        this.tail = new_node
        this.length = 1
    }

    printList() {
        let temp = this.head
        for (let i = 0; i < this.length; i++) {
            console.log(temp.value)
            temp = temp.next
        }
    }

    getValueAtIndex(index) {
        if (index < 0 || index >= this.length) {
            return false
        }
        let temp = this.head
        let count = 0;
        while (count < index) {
            temp = temp.next
            count ++
        }
        return temp
    }

    append(value) {
        let new_node = new Node(value)
        if (this.length == 0) {
            this.head = new_node
            this.tail = new_node
        }
        else {
            this.tail.next = new_node
            this.tail = new_node
        }
        this.length ++
        return true
    } 
    prepend(value) {
        let new_node = new Node(value)
        if (this.length == 0) {
            this.head = new_node
            this.tail = new_node
        }
        else {
            new_node.next = this.head
            this.head = new_node
        }
        this.length ++
        return true
    }

    pop() {
        if (this.length == 0) {
            return null
        }
        let temp = this.head
        let pre = this.tail

        while (temp.next) {
            pre = temp
            temp = temp.next
        }

        this.tail = pre
        this.tail.next = null
        this.length -- 
        return temp
    } 

    contains(value) {
        let temp = this.head
        for (let i = 0; i < this.length; i++) {
            if (temp.value == value) {
                return `Value ${value} at index ${i}`
            }
            temp = temp.next
        }
        return `Value ${value} not in linked list`
    }

    
    insert(index, value) {
        if (index < 0 || index > this.length) {
            return false
        }
        if (index == 0) {
            return this.prepend(value)
        }
        if (index == this.length) {
            return this.append(value)
        }
        let new_node = new Node(value)
        let temp = this.getValueAtIndex(index - 1)
        new_node.next = temp.next
        temp.next = new_node
        this.length ++
    }
        
}

// Test Cases

let coolList = new LinkedList(1)

coolList.append(2)
coolList.append(3)
coolList.append(4)
coolList.append(5)
coolList.append(6)

coolList.prepend(0)

coolList.printList()

console.log(`Linked List contains ${coolList.length} nodes`)

console.log(coolList.contains(7))

console.log(coolList.contains(3))

coolList.insert(3, 'dance party!!!!')

coolList.insert(8, 7)

coolList.insert(7, 'another dance party!!!!!!')

coolList.printList()

File: dsa/linked-list/linked-list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

# LL Constructor w/ one node
class LinkedList:
    def __init__(self, value):
        new_node = Node(value)
        self.head = new_node
        self.tail = new_node
        self.length = 1
    
    def print_list(self):
        temp = self.head
        while temp:
            print(temp.value)
            temp = temp.next

    def append(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else: 
            # Add new node to end of linked list
            self.tail.next = new_node
            # Set tail to newly added node
            self.tail = new_node
        self.length += 1
        # Used for methods that call append() and require boolean
        return True
    
    def pop(self):
        if self.length == 0:
            return None
        temp = self.head
        pre = self.head

        while temp.next:
            pre = temp
            # Move temp pointer to the end of the linked list
            temp = temp.next
        # Pre will be pointing at second to last node
        # Set new tail to node pre is pointing to
        self.tail = pre
        # Pop end node off linked list
        self.tail.next = None
        self.length -= 1
        # If the linked list only had one node, use this to reset head and tail pointers
        if self.length == 0:
            self.head = None
            self.tail = None
        return temp

    def prepend(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            # Point new node to head
            new_node.next = self.head
            # Set head to newly prepended node
            self.head = new_node
        self.length += 1
        # Used for methods that call append() and require boolean
        return True

    def pop_first(self):
        if self.length == 0:
            return None
        # Set temp variable to first node in list
        temp = self.head
        # Set head pointer to next Node
        self.head = self.head.next
        # Remove first node from list
        temp.next = None
        self.length -= 1
        # When starting w/ one item
        if self.length == 0:
            self.tail = None
        return temp

    # Return value of item in index
    def get(self, index):
        # Prevent out of range error
        if index < 0 or index >= self.length:
            return None
        temp = self.head
        # Iterate to the length of the index
        # Use _ as it is not being used inside for loop
        for _ in range(index):
            temp = temp.next
        return temp

    # Set value at given index (think change value - not insert)
    def set_value(self, index, value):
        temp = self.get(index)
        if temp: 
            temp.value = value
            return True
        # If temp value returns None
        return False


    def insert(self, index, value):
        # Out of range
        if index < 0 or index > self.length:
            return False
        # Add node to beginning 
        if index == 0:
            return self.prepend(value)
        # Add node to end
        if index == self.length:
            return self.append(value)
        # Insert new node somewhere else in list
        new_node = Node(value)
        # Point temp to given node
        temp = self.get(index - 1)
        # Point new node to next node
        new_node.next = temp.next
        # Point previous node at index to new node
        temp.next = new_node
        self.length += 1
        return True 


    def remove(self, index):
        

    #def reverse():


## Time Complexity for each Method

## Space Complexity

## Test Methods ## 

my_linked_list = LinkedList(7)


my_linked_list.append(9)
my_linked_list.append(2)
my_linked_list.append(13)

print(f'Removed node {my_linked_list.pop().value}')

my_linked_list.prepend(12)

my_linked_list.pop_first()

# Return value of item in index 1
print(f'The value at index 1: {my_linked_list.get(1).value}')


my_linked_list.insert(2, 27)

my_linked_list.print_list()


